plugins/module_utils/data_set.py

```diff
@@ -1025,8 +1025,8 @@ def create(
         except exceptions.ZOAUException as create_exception:
             raise DatasetCreateError(
                 name,
-                create_exception.response.rc,
-                create_exception.response.stdout_response + create_exception.response.stderr_response
+                create_exception.rc,
+                create_exception.stdout_response + create_exception.stderr_response
```

Comment:
Curious what triggered this change, from the datasets.py code (internal ZOAU Python repo):
```

# Exceptions from ZOAU Core Calls
class _ZOAUExtendableException(Exception):
    """Custom exception supporting a string message and ZOAUResponse object
    Attributes
    ----------
    message: str
        ZOAUResponse as string (Printed as output by `raise`)
    response : ZOAUResponse
        Addressable ZOAUResponse object
    Notes
    -----
    `ZOAUResponse` attributes
        - rc: str
            Return code
        - stdout_response: str
            Stdout
        - stderr_response: str
            Stderr
        - command: str
            Command performed during a ZOAU core call
        - response_format: str
            Response encoding
    """

    message: str
    response: ZOAUResponse

    def __init__(self, response_obj: ZOAUResponse):
        super().__init__(str(response_obj))
        self.response = response_obj


class ZOAUException(_ZOAUExtendableException):
    """Errors happening during ZOAU core calls, see message for details"""

```
It appears that all the response data (`rc`, `response_stdout`, `response_stderr`) are inside of a `ZOAUResponse` object, which is folded into the `self.response`. So I would think that getting at those would require the `create_exception.response.<item>` where the `self` in this block would be the Exception variable `create_exception`.

In a (very) rough re-create, I also found that we should be `except`-ing a `_ZOAUExtendableException` rather than a `ZOAUException` as ZOAU's internal exception inheritance seems to have undergone a re-structuring where many of the Exceptions now seem to extend the `_ZOAUExtendableException`. 

(This should have ideally been a part of the module migration work, but it appears it was missed, oops) -- I can open up a separate work item to encapsulate/track all this if the changes and validation seem large.

The recreate induces a `DatasetWriteException` which is not triggered from the `datasets.create()` function, but I found it simpler to trigger this error to play around with rather than an error from the create function. I figure they're similar enough since they both have similar definitions and both extend the same `_ZOAUExtendableException` class.

Rough re-create:
```python3
>>> from zoautil_py import datasets, exceptions
>>> 
>>> datasets.write("some.data.set", "helllothere")                                        
0
```
The above should create and write a line to the data set. From here, I opened up the data set in ISPF (`=3.4`) and browsed (`b`) it. Without forcing the write, I could trigger a `DatasetWriteException`:
```python3
>>> try:                                                           
...     datasets.write("some.data.set", "helllothere")             
... except exceptions.ZOAUException as create_exception: 
...     print('hello from ZOAUException')
... except:
...     print('something else')
... 
something else
``` 

This suggested to me that we're capturing the wrong exception. Trying with 
```python3
>>> try:                                                           
...     datasets.write("some.data.set", "helllothere")             
... except exceptions._ZOAUExtendableException as create_exception: 
...     print('hello from ZOAUException')
... except:
...     print('something else')
... 
hello from ZOAUException
``` 

And finally, the contents of the exception:
```python3
>>> try:                                                           
...     datasets.write("some.data.set", "helllothere")             
... except exceptions._ZOAUExtendableException as create_exception:
...     st = str(create_exception.rc) + create_exception.stderr_response
...     print('hi', st)                                            
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/zoau/v1.3.0/lib/3.11/zoautil_py/datasets.py", line 554, in write
    raise DatasetWriteException(response)
zoautil_py.exceptions.DatasetWriteException: [ZOAUResponse]
	rc: 8
	response_format: UTF-8
	stdout_response: 
	stderr_response: BGYSC0705E Failed to open dataset SOME.DATA.SET for write.

	command: decho -- "helllothere" "some.data.set"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
AttributeError: 'DatasetWriteException' object has no attribute 'rc'
>>> 
```
The final line above suggested to me that the `rc`, `stdout` and `stderr` are nested in the `response` attribute:
```python3
>>> try:                                                                                                                              
...     datasets.write("some.data.set", "helllothere")                                                                                
... except exceptions._ZOAUExtendableException as create_exception:                                                                   
...     st = str(create_exception.response.rc) + create_exception.response.stderr_response
...     print('hi', st)                                                                                                               
... 
hi 8BGYSC0705E Failed to open dataset SOME.DATA.SET for write.
```

Reply:
There are 2 exceptions in that segment of code: exceptions.ZOAUException, and exception.DatasetCreationError.  Where I'm a little confused is how to make this better: ZOAUException does not pass .response... I can switch the first on to extendable... where I'm uncertain is should this also apply to DatasetCreationError: should it just go away, stay as is, or what?  I could use some ideas on this one.

Reply:
I think the code should become:
```python3
  <...>
        try:
            datasets.create(**formatted_args)
        except exceptions._ZOAUExtendableException as create_exception:
            raise DatasetCreateError(
                name,
                create_exception.response.rc,
                create_exception.response.stdout_response + create_exception.response.stderr_response
            )
        except exceptions.DatasetVerificationError as e:
            # verification of a data set spanning multiple volumes is currently broken in ZOAU v.1.3.0
            if volumes and len(volumes) > 1:
                if DataSet.data_set_cataloged(name, volumes):
                    return 0
            raise DatasetCreateError(
                name,
                msg="Unable to verify the data set was created. Received DatasetVerificationError from ZOAU.",
            )
        # With ZOAU 1.3 we switched from getting a ZOAUResponse obj to a Dataset obj, previously we returned
        # response.rc now we just return 0 if nothing failed
        return 0
```

The way I see it, `datasets.create()` can do 1 of 3 things:
1. succeed (yay!)
2. throw an exception which extends `_ZOAUExtendableException`
3. throw an exception which does not extend `_ZOAUExtendableException` merely extends the built-in base `Exception` class

scenario 1 is straightforward.

Scenarios 2 and 3 are non-obvious because in ZOAU python API documentation, it is not mentioned that these *two* distinct categories of exceptions exist. They even state in their docs that all their exceptions extend from `ZOAUException`, which isn't exactly right as of right now.
![image](https://github.com/ansible-collections/ibm_zos_core/assets/10567394/38fc2aea-14d6-4189-8b02-27992f67f58b)


My understanding of the reasoning behind their decision is that they're still determining what information to use to construct those exceptions which do not `_ZOAUExtendableException`. 

I've been referencing the code in their IBM internal repo for their python API -- specifically the `exceptions.py` file, where the exceptions are defined:

```python3
class DatasetCreateException(_ZOAUExtendableException):
    """Errors while trying to create the dataset, see message for details"""
```
```python3
class DatasetVerificationError(Exception):
    pass
```
The `datasets.create()` function we (Ansible) call may raise either of these two exceptions and each gets handled differently.

Also, mainly because I just got slightly dizzy thinking about it, `DatasetCreateException` is defined by ZOAU and `DatasetCreateError` is defined by us (Ansible).

We `raise` a `DatasetCreateError` which is a formatted string `msg`, which we populate based on what we receive from the call to ZOAU via `datasets.create()`. What we receive could be `rc`/`stdout`/`stderr` packaged inside a ZOAUResponse object if the exception extends `_ZOAUExtendableException` or we could get an error message string if we catch something like a `DatasetVerificationError` which just sends out the data set name.

I somewhat recently modified the definition of `DatasetCreateError` so that the string we're constructing may or may not include an RC. This way, for a `DatasetVerificationError`, we do not have to "make up" our own RC and still get the message out to the user.

Reply:
I think I understand where the disconnect was.  I'll re-run the new code through the pipeline.

Reply:
Great analysis Ketan, it was unclear for me too and now I get it